### PR TITLE
Make use of CMake 3.22 features

### PR DESCRIPTION
**Pull request by alexreinking**
Now that we have CMake 3.22 as a minimum version, we might as well use its features. This PR is going to do this in one fell swoop. I have:

* replaced our `add_cuda_to_target` helper with the [`CUDAToolkit`](https://cmake.org/cmake/help/latest/module/FindCUDAToolkit.html) module (CMake 3.17+)
* modernized `target_export_script` to use the [`CheckLinkerFlag`](https://cmake.org/cmake/help/latest/module/CheckLinkerFlag.html) module (CMake 3.18+)
* made `find_program` and `find_library` use [`REQUIRED`](https://cmake.org/cmake/help/latest/command/find_library.html) where appropriate. (CMake 3.18+)
* reworked the linear_algebra app's BLAS detection. Uses [`BLAS::BLAS`](https://cmake.org/cmake/help/latest/module/FindBLAS.html) target for default instead of custom logic. (CMake 3.18+)
* changed uses of `get_filename_component` to [`cmake_path`](https://cmake.org/cmake/help/latest/command/cmake_path.html), which superseded it (CMake 3.20+)
* simplified `src/runtime/CMakeLists.txt` since [`DEPFILE`](https://cmake.org/cmake/help/latest/command/add_custom_command.html#generating-files) works everywhere now (CMake 3.21+)
* use [sane cache behavior](https://cmake.org/cmake/help/latest/policy/CMP0126.html) in HalideTargetHelpers.cmake (CMake 3.21+)
* replaced our `make_shell_path` helper with the [`ENVIRONMENT_MODIFICATION`](https://cmake.org/cmake/help/latest/prop_test/ENVIRONMENT_MODIFICATION.html) test property (CMake 3.22+)

This is all for now. Changes to the Python build are to-do in a separate PR.

### CODE REVIEW

File: test/generator/CMakeLists.txt [hunk 1]
```diff
@@ -239,23 +241,23 @@ halide_define_aot_test(float16_t)
...
-  add_halide_library(gpu_multi_context_threaded_mul FROM gpu_multi_context_threaded.generator
-                     FEATURES user_context)
-
-  if (TARGET_NVPTX AND Halide_TARGET MATCHES "cuda")
-      add_cuda_to_target(generator_aot_gpu_multi_context_threaded PRIVATE)
-  endif ()
-  if (TARGET_NVPTX AND Halide_TARGET MATCHES "opencl")
-      target_link_libraries(generator_aot_gpu_multi_context_threaded PRIVATE OpenCL::OpenCL)
-  endif ()
+    halide_define_aot_test(gpu_multi_context_threaded
```
Comment (alexreinking) [inline test/generator/CMakeLists.txt:244]: Indentation was inconsistent in this part of the file...

**Comment by steven-johnson** [comment]
BLAS detection code is now broken on OSX: https://buildbot.halide-lang.org/master/#/builders/86/builds/50

**Comment by steven-johnson** [comment]
And also WIndows: https://buildbot.halide-lang.org/master/#/builders/10/builds/49

**Comment by steven-johnson** [comment]
Windows still broken, alas: https://buildbot.halide-lang.org/master/#/builders/10/builds/50

**Review by steven-johnson — APPROVED** [review]
LGTM pending all-green

**Comment by alexreinking** [comment]
> Windows still broken, alas

Resolution: our buildbots hadn't updated vcpkg since 2020, which is older than CMake 3.22 (Nov 18, 2021). Upgrading vcpkg resolved this issue.

**Pull request merged**